README.rst
===============================
approx
===============================

.. image:: https://img.shields.io/travis/jeroyang/approx.svg
        :target: https://travis-ci.org/jeroyang/approx

.. image:: https://img.shields.io/pypi/v/approx.svg
        :target: https://pypi.python.org/pypi/approx


A game for mental mathematicsin approximate numbers

* Free software: BSD license
* Documentation: https://approx.readthedocs.org.

Features
--------

* TODO


HISTORY.rst
.. :changelog:

History
-------

 (2015-11-30)
---------------------

* First release on PyPI.


tests/test_approx.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from __future__ import (absolute_import, division,
                        print_function, unicode_literals)
from builtins import *

"""
test_approx
----------------------------------

Tests for `approx` module.
"""

import unittest

from approx import approx

class TestApprox(unittest.TestCase):
    def test_question_guess(self):
        question, answer = approx.question_guess(5, 1, 100)
        wanted_question = '從 1 到 100 猜一個數字：'
        wanted_answer = 5
        self.assertEqual(question, wanted_question)
        self.assertEqual(answer, wanted_answer)

    def test_question_add(self):
        question, answer = approx.question_add(3, 5)
        wanted_question = '3 + 5 = '
        wanted_answer = 8
        self.assertEqual(question, wanted_question)
        self.assertEqual(answer, wanted_answer)

    def test_question_sub(self):
        question, answer = approx.question_sub(8, 7)
        wanted_question = '15 - 8 = '
        wanted_answer = 7
        self.assertEqual(question, wanted_question)
        self.assertEqual(answer, wanted_answer)

    def test_question_multiply(self):
        question, answer = approx.question_multiply(3, 5)
        wanted_question = '3 × 5 = '
        wanted_answer = 15
        self.assertEqual(question, wanted_question)
        self.assertEqual(answer, wanted_answer)

    def test_question_divide(self):
        question, answer = approx.question_divide(3, 5)
        wanted_question = '15 ÷ 3 = '
        wanted_answer = 5
        self.assertEqual(question, wanted_question)
        self.assertEqual(answer, wanted_answer)

    def test_question_highest_digit(self):
        args = ('15 + 16 = ', 31)
        question, answer = approx.question_highest_digit(*args)
        wanted_question = '15 + 16 = ▢1，其中▢應填入什麼數字？'
        wanted_answer = 3
        self.assertEqual(question, wanted_question)
        self.assertEqual(answer, wanted_answer)
        args = ('300 - 120 = ', 180)
        question, answer = approx.question_highest_digit(*args)
        wanted_question = '300 - 120 = ▢80，其中▢應填入什麼數字？'
        wanted_answer = 1
        self.assertEqual(question, wanted_question)
        self.assertEqual(answer, wanted_answer)

    def test_is_correct(self):
        correctness = approx.is_correct(800, 1000, 100)
        self.assertEqual(correctness, False)

        correctness = approx.is_correct(1055, 1000, 100)
        self.assertEqual(correctness, True)

        correctness = approx.is_correct(1055, 1000, 50)
        self.assertEqual(correctness, False)

    def test_new_level_greet(self):
        result = approx.new_level_greet(11, 0)
        bar = '=' * 20
        wanted = '\n'.join([bar, '第 11 關 (容許誤差：0)', bar])
        self.assertEqual(result, wanted)

    def test_correct_greet(self):
        result = approx.correct_greet(10, 11)
        wanted = '算你答對，正確答案是 10！'
        self.assertEqual(result, wanted)
        result = approx.correct_greet(10, 10)
        wanted = '太棒了，答案就是 10！'
        self.assertEqual(result, wanted)

    def test_too_high_hint(self):
        result = approx.too_high_hint()
        wanted = '太多了，再少一點！'
        self.assertEqual(result, wanted)

    def test_too_low_hint(self):
        result = approx.too_low_hint()
        wanted = '太少了，再多一點！'
        self.assertEqual(result, wanted)

    def test_generate_question(self):
        result = approx.generate_question('add', 1, 3)
        wanted = ('1 + 3 = ', 4)
        self.assertEqual(result, wanted)

    def test_play_question(self):
        question = 'anything question'
        answer = 15
        precision = 0
        out = []
        def print_method(text):
            out.append(text)

        input_sequence = iter(['13', '15'])
        def input_func(text):
            return next(input_sequence)

        result = approx.play_question(
            question, answer, precision, print_method, input_func)
        wanted = True
        self.assertEqual(result, wanted)
        self.assertEqual(out, ['太少了，再多一點！', '太棒了，答案就是 15！'])

    def test_confirm_exit(self):
        result = approx.confirm_exit(lambda x: '1')
        self.assertTrue(result)
        result = approx.confirm_exit(lambda x: 'd')
        self.assertFalse(result)

class TestApproxGame(unittest.TestCase):

    def setUp(self):
        self.game = approx.ApproxGame()

    def tearDown(self):
        pass


approx/_version.py

# This file helps to compute a version number in source trees obtained from
# git-archive tarball (such as those provided by githubs download-from-tag
# feature). Distribution tarballs (built by setup.py sdist) and build
# directories (produced by setup.py build) will contain a much shorter file
# that just contains the computed version number.

# This file is released into the public domain. Generated by
# versioneer-0.15 (https://github.com/warner/python-versioneer)

import errno
import os
import re
import subprocess
import sys


def get_keywords():
    # these strings will be replaced by git during git-archive.
    # setup.py/versioneer.py will grep for the variable names, so they must
    # each be defined on a line of their own. _version.py will just call
    # get_keywords().
    git_refnames = "$Format:%d$"
    git_full = "$Format:%H$"
    keywords = {"refnames": git_refnames, "full": git_full}
    return keywords


class VersioneerConfig:
    pass


def get_config():
    # these strings are filled in when 'setup.py versioneer' creates
    # _version.py
    cfg = VersioneerConfig()
    cfg.VCS = "git"
    cfg.style = "pep440"
    cfg.tag_prefix = ""
    cfg.parentdir_prefix = ""
    cfg.versionfile_source = "approx/_version.py"
    cfg.verbose = False
    return cfg


class NotThisMethod(Exception):
    pass


LONG_VERSION_PY = {}
HANDLERS = {}


def register_vcs_handler(vcs, method):  # decorator
    def decorate(f):
        if vcs not in HANDLERS:
            HANDLERS[vcs] = {}
        HANDLERS[vcs][method] = f
        return f
    return decorate


def run_command(commands, args, cwd=None, verbose=False, hide_stderr=False):
    assert isinstance(commands, list)
    p = None
    for c in commands:
        try:
            dispcmd = str([c] + args)
            # remember shell=False, so use git.cmd on windows, not just git
            p = subprocess.Popen([c] + args, cwd=cwd, stdout=subprocess.PIPE,
                                 stderr=(subprocess.PIPE if hide_stderr
                                         else None))
            break
        except EnvironmentError:
            e = sys.exc_info()[1]
            if e.errno == errno.ENOENT:
                continue
            if verbose:
                print("unable to run %s" % dispcmd)
                print(e)
            return None
    else:
        if verbose:
            print("unable to find command, tried %s" % (commands,))
        return None
    stdout = p.communicate()[0].strip()
    if sys.version_info[0] >= 3:
        stdout = stdout.decode()
    if p.returncode != 0:
        if verbose:
            print("unable to run %s (error)" % dispcmd)
        return None
    return stdout


def versions_from_parentdir(parentdir_prefix, root, verbose):
    # Source tarballs conventionally unpack into a directory that includes
    # both the project name and a version string.
    dirname = os.path.basename(root)
    if not dirname.startswith(parentdir_prefix):
        if verbose:
            print("guessing rootdir is '%s', but '%s' doesn't start with "
                  "prefix '%s'" % (root, dirname, parentdir_prefix))
        raise NotThisMethod("rootdir doesn't start with parentdir_prefix")
    return {"version": dirname[len(parentdir_prefix):],
            "full-revisionid": None,
            "dirty": False, "error": None}


@register_vcs_handler("git", "get_keywords")
def git_get_keywords(versionfile_abs):
    # the code embedded in _version.py can just fetch the value of these
    # keywords. When used from setup.py, we don't want to import _version.py,
    # so we do it with a regexp instead. This function is not used from
    # _version.py.
    keywords = {}
    try:
        f = open(versionfile_abs, "r")
        for line in f.readlines():
            if line.strip().startswith("git_refnames ="):
                mo = re.search(r'=\s*"(.*)"', line)
                if mo:
                    keywords["refnames"] = mo.group(1)
            if line.strip().startswith("git_full ="):
                mo = re.search(r'=\s*"(.*)"', line)
                if mo:
                    keywords["full"] = mo.group(1)
        f.close()
    except EnvironmentError:
        pass
    return keywords


@register_vcs_handler("git", "keywords")
def git_versions_from_keywords(keywords, tag_prefix, verbose):
    if not keywords:
        raise NotThisMethod("no keywords at all, weird")
    refnames = keywords["refnames"].strip()
    if refnames.startswith("$Format"):
        if verbose:
            print("keywords are unexpanded, not using")
        raise NotThisMethod("unexpanded keywords, not a git-archive tarball")
    refs = set([r.strip() for r in refnames.strip("()").split(",")])
    # starting in git-1.8.3, tags are listed as "tag: foo-1.0" instead of
    # just "foo-1.0". If we see a "tag: " prefix, prefer those.
    TAG = "tag: "
    tags = set([r[len(TAG):] for r in refs if r.startswith(TAG)])
    if not tags:
        # Either we're using git < 1.8.3, or there really are no tags. We use
        # a heuristic: assume all version tags have a digit. The old git %d
        # expansion behaves like git log --decorate=short and strips out the
        # refs/heads/ and refs/tags/ prefixes that would let us distinguish
        # between branches and tags. By ignoring refnames without digits, we
        # filter out many common branch names like "release" and
        # "stabilization", as well as "HEAD" and "master".
        tags = set([r for r in refs if re.search(r'\d', r)])
        if verbose:
            print("discarding '%s', no digits" % ",".join(refs-tags))
    if verbose:
        print("likely tags: %s" % ",".join(sorted(tags)))
    for ref in sorted(tags):
        # sorting will prefer e.g. "2.0" over "2.0rc1"
        if ref.startswith(tag_prefix):
            r = ref[len(tag_prefix):]
            if verbose:
                print("picking %s" % r)
            return {"version": r,
                    "full-revisionid": keywords["full"].strip(),
                    "dirty": False, "error": None
                    }
    # no suitable tags, so version is "0+unknown", but full hex is still there
    if verbose:
        print("no suitable tags, using unknown + full revision id")
    return {"version": "0+unknown",
            "full-revisionid": keywords["full"].strip(),
            "dirty": False, "error": "no suitable tags"}


@register_vcs_handler("git", "pieces_from_vcs")
def git_pieces_from_vcs(tag_prefix, root, verbose, run_command=run_command):
    # this runs 'git' from the root of the source tree. This only gets called
    # if the git-archive 'subst' keywords were *not* expanded, and
    # _version.py hasn't already been rewritten with a short version string,
    # meaning we're inside a checked out source tree.

    if not os.path.exists(os.path.join(root, ".git")):
        if verbose:
            print("no .git in %s" % root)
        raise NotThisMethod("no .git directory")

    GITS = ["git"]
    if sys.platform == "win32":
        GITS = ["git.cmd", "git.exe"]
    # if there is a tag, this yields TAG-NUM-gHEX[-dirty]
    # if there are no tags, this yields HEX[-dirty] (no NUM)
    describe_out = run_command(GITS, ["describe", "--tags", "--dirty",
                                      "--always", "--long"],
                               cwd=root)
    # --long was added in git-1.5.5
    if describe_out is None:
        raise NotThisMethod("'git describe' failed")
    describe_out = describe_out.strip()
    full_out = run_command(GITS, ["rev-parse", "HEAD"], cwd=root)
    if full_out is None:
        raise NotThisMethod("'git rev-parse' failed")
    full_out = full_out.strip()

    pieces = {}
    pieces["long"] = full_out
    pieces["short"] = full_out[:7]  # maybe improved later
    pieces["error"] = None

    # parse describe_out. It will be like TAG-NUM-gHEX[-dirty] or HEX[-dirty]
    # TAG might have hyphens.
    git_describe = describe_out

    # look for -dirty suffix
    dirty = git_describe.endswith("-dirty")
    pieces["dirty"] = dirty
    if dirty:
        git_describe = git_describe[:git_describe.rindex("-dirty")]

    # now we have TAG-NUM-gHEX or HEX

    if "-" in git_describe:
        # TAG-NUM-gHEX
        mo = re.search(r'^(.+)-(\d+)-g([0-9a-f]+)$', git_describe)
        if not mo:
            # unparseable. Maybe git-describe is misbehaving?
            pieces["error"] = ("unable to parse git-describe output: '%s'"
                               % describe_out)
            return pieces

        # tag
        full_tag = mo.group(1)
        if not full_tag.startswith(tag_prefix):
            if verbose:
                fmt = "tag '%s' doesn't start with prefix '%s'"
                print(fmt % (full_tag, tag_prefix))
            pieces["error"] = ("tag '%s' doesn't start with prefix '%s'"
                               % (full_tag, tag_prefix))
            return pieces
        pieces["closest-tag"] = full_tag[len(tag_prefix):]

        # distance: number of commits since tag
        pieces["distance"] = int(mo.group(2))

        # commit: short hex revision ID
        pieces["short"] = mo.group(3)

    else:
        # HEX: no tags
        pieces["closest-tag"] = None
        count_out = run_command(GITS, ["rev-list", "HEAD", "--count"],
                                cwd=root)
        pieces["distance"] = int(count_out)  # total number of commits

    return pieces


def plus_or_dot(pieces):
    if "+" in pieces.get("closest-tag", ""):
        return "."
    return "+"


def render_pep440(pieces):
    # now build up version string, with post-release "local version
    # identifier". Our goal: TAG[+DISTANCE.gHEX[.dirty]] . Note that if you
    # get a tagged build and then dirty it, you'll get TAG+0.gHEX.dirty

    # exceptions:
    # 1: no tags. git_describe was just HEX. 0+untagged.DISTANCE.gHEX[.dirty]

    if pieces["closest-tag"]:
        rendered = pieces["closest-tag"]
        if pieces["distance"] or pieces["dirty"]:
            rendered += plus_or_dot(pieces)
            rendered += "%d.g%s" % (pieces["distance"], pieces["short"])
            if pieces["dirty"]:
                rendered += ".dirty"
    else:
        # exception #1
        rendered = "0+untagged.%d.g%s" % (pieces["distance"],
                                          pieces["short"])
        if pieces["dirty"]:
            rendered += ".dirty"
    return rendered


def render_pep440_pre(pieces):
    # TAG[.post.devDISTANCE] . No -dirty

    # exceptions:
    # 1: no tags. 0.post.devDISTANCE

    if pieces["closest-tag"]:
        rendered = pieces["closest-tag"]
        if pieces["distance"]:
            rendered += ".post.dev%d" % pieces["distance"]
    else:
        # exception #1
        rendered = "0.post.dev%d" % pieces["distance"]
    return rendered


def render_pep440_post(pieces):
    # TAG[.postDISTANCE[.dev0]+gHEX] . The ".dev0" means dirty. Note that
    # .dev0 sorts backwards (a dirty tree will appear "older" than the
    # corresponding clean one), but you shouldn't be releasing software with
    # -dirty anyways.

    # exceptions:
    # 1: no tags. 0.postDISTANCE[.dev0]

    if pieces["closest-tag"]:
        rendered = pieces["closest-tag"]
        if pieces["distance"] or pieces["dirty"]:
            rendered += ".post%d" % pieces["distance"]
            if pieces["dirty"]:
                rendered += ".dev0"
            rendered += plus_or_dot(pieces)
            rendered += "g%s" % pieces["short"]
    else:
        # exception #1
        rendered = "0.post%d" % pieces["distance"]
        if pieces["dirty"]:
            rendered += ".dev0"
        rendered += "+g%s" % pieces["short"]
    return rendered


def render_pep440_old(pieces):
    # TAG[.postDISTANCE[.dev0]] . The ".dev0" means dirty.

    # exceptions:
    # 1: no tags. 0.postDISTANCE[.dev0]

    if pieces["closest-tag"]:
        rendered = pieces["closest-tag"]
        if pieces["distance"] or pieces["dirty"]:
            rendered += ".post%d" % pieces["distance"]
            if pieces["dirty"]:
                rendered += ".dev0"
    else:
        # exception #1
        rendered = "0.post%d" % pieces["distance"]
        if pieces["dirty"]:
            rendered += ".dev0"
    return rendered


def render_git_describe(pieces):
    # TAG[-DISTANCE-gHEX][-dirty], like 'git describe --tags --dirty
    # --always'

    # exceptions:
    # 1: no tags. HEX[-dirty]  (note: no 'g' prefix)

    if pieces["closest-tag"]:
        rendered = pieces["closest-tag"]
        if pieces["distance"]:
            rendered += "-%d-g%s" % (pieces["distance"], pieces["short"])
    else:
        # exception #1
        rendered = pieces["short"]
    if pieces["dirty"]:
        rendered += "-dirty"
    return rendered


def render_git_describe_long(pieces):
    # TAG-DISTANCE-gHEX[-dirty], like 'git describe --tags --dirty
    # --always -long'. The distance/hash is unconditional.

    # exceptions:
    # 1: no tags. HEX[-dirty]  (note: no 'g' prefix)

    if pieces["closest-tag"]:
        rendered = pieces["closest-tag"]
        rendered += "-%d-g%s" % (pieces["distance"], pieces["short"])
    else:
        # exception #1
        rendered = pieces["short"]
    if pieces["dirty"]:
        rendered += "-dirty"
    return rendered


def render(pieces, style):
    if pieces["error"]:
        return {"version": "unknown",
                "full-revisionid": pieces.get("long"),
                "dirty": None,
                "error": pieces["error"]}

    if not style or style == "default":
        style = "pep440"  # the default

    if style == "pep440":
        rendered = render_pep440(pieces)
    elif style == "pep440-pre":
        rendered = render_pep440_pre(pieces)
    elif style == "pep440-post":
        rendered = render_pep440_post(pieces)
    elif style == "pep440-old":
        rendered = render_pep440_old(pieces)
    elif style == "git-describe":
        rendered = render_git_describe(pieces)
    elif style == "git-describe-long":
        rendered = render_git_describe_long(pieces)
    else:
        raise ValueError("unknown style '%s'" % style)

    return {"version": rendered, "full-revisionid": pieces["long"],
            "dirty": pieces["dirty"], "error": None}


def get_versions():
    # I am in _version.py, which lives at ROOT/VERSIONFILE_SOURCE. If we have
    # __file__, we can work backwards from there to the root. Some
    # py2exe/bbfreeze/non-CPython implementations don't do __file__, in which
    # case we can only use expanded keywords.

    cfg = get_config()
    verbose = cfg.verbose

    try:
        return git_versions_from_keywords(get_keywords(), cfg.tag_prefix,
                                          verbose)
    except NotThisMethod:
        pass

    try:
        root = os.path.realpath(__file__)
        # versionfile_source is the relative path from the top of the source
        # tree (where the .git directory might live) to this file. Invert
        # this to find the root from __file__.
        for i in cfg.versionfile_source.split('/'):
            root = os.path.dirname(root)
    except NameError:
        return {"version": "0+unknown", "full-revisionid": None,
                "dirty": None,
                "error": "unable to find root of source tree"}

    try:
        pieces = git_pieces_from_vcs(cfg.tag_prefix, root, verbose)
        return render(pieces, cfg.style)
    except NotThisMethod:
        pass

    try:
        if cfg.parentdir_prefix:
            return versions_from_parentdir(cfg.parentdir_prefix, root, verbose)
    except NotThisMethod:
        pass

    return {"version": "0+unknown", "full-revisionid": None,
            "dirty": None,
            "error": "unable to compute version"}


approx/approx.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from __future__ import (absolute_import, division,
                        print_function, unicode_literals)
from builtins import *
from random import choice

def direct_print(text):
    print(text, flush=True)
    
def question_guess(number, lower_bound, upper_bound):
    template = '從 {} 到 {} 猜一個數字：'
    question = template.format(lower_bound, upper_bound)
    answer = number
    return (question, answer)

def question_add(arg1, arg2):
    template = '{0} + {1} = '
    question = template.format(arg1, arg2)
    answer = arg1 + arg2
    return (question, answer)

def question_sub(arg1, wanted_answer):
    template = '{0} - {1} = '
    question = template.format(arg1 + wanted_answer, arg1)
    answer = wanted_answer
    return (question, answer)

def question_sub_fixed(arg1, arg2):
    """
    Given arg1 and arg2 return the question of "arg1 - arg2 = "
    """
    template = '{0} - {1} = '
    question = template.format(arg1, arg2)
    answer = arg1 - arg2
    return (question, answer)

def question_multiply(arg1, arg2):
    template = '{0} × {1} = '
    question = template.format(arg1, arg2)
    answer = arg1 * arg2
    return (question, answer)

def question_divide(arg1, wanted_answer):
    template = '{0} ÷ {1} = '
    question = template.format(arg1*wanted_answer, arg1)
    answer = wanted_answer
    return (question, answer)

def question_highest_digit(original_question, original_answer):
    blocked_answer = '▢' + str(original_answer)[1:]
    question = ''.join([
        original_question,
        blocked_answer,
        '，其中▢應填入什麼數字？',
        ])
    answer = int(str(original_answer)[0])
    return (question, answer)

def question_highest_wrapper(recipe, *args):
    original = generate_question(recipe, *args)
    question, answer = question_highest_digit(*original)
    return (question, answer)

def is_correct(users_answer, answer, precision):
    delta = abs(users_answer - answer)
    
    return delta <= precision

def new_level_greet(level_id, precision):
    template = '第 {} 關 (容許誤差：{})'
    greet = template.format(level_id, precision)
    bar = '=' * 20
    return '\n'.join([bar, greet, bar])

def correct_greet(answer, users_answer):
    if answer == users_answer:
        greet = '太棒了，答案就是 {}！'.format(answer)
    else:
        greet = '算你答對，正確答案是 {}！'.format(answer)
    return greet

def too_high_hint():
    return '太多了，再少一點！'

def too_low_hint():
    return '太少了，再多一點！'
    
RECIPE_MAP = {
    'guess': question_guess,
    'add': question_add,
    'sub': question_sub,
    'subf': question_sub_fixed,
    'multiply': question_multiply,
    'divide': question_divide,
    'highest': question_highest_wrapper,
}

def random_from_range(*args):
    """
    Return corresponding args, choose one number from given range
    """
    for arg in args:
        if isinstance(arg, range):
            yield choice(arg)
        else: 
            yield arg

def generate_question(recipe, *args):
    """
    The recipe is a key in the RECIPE_MAP,

    """
    question_fucntion = RECIPE_MAP[recipe]
    randomized_args = random_from_range(*args)
    question, answer = question_fucntion(*randomized_args)
    return question, answer

def play_question(question, answer, precision, print_method, input_func):
    while True:
        users_response = input_func(question)
        try:
            users_answer = float(users_response)
        except:
            confirmed = confirm_exit(input_func)
            if confirmed:
                return False
            else:
                continue
        
        if is_correct(users_answer, answer, precision):
            print_method(correct_greet(answer, users_answer))
            return True
        
        elif users_answer > answer:
            print_method(too_high_hint())
        else: #users_answer < answer
            print_method(too_low_hint())

def confirm_exit(input_func):
    answer = input_func("確認結束遊戲？(是請按1；其他鍵表示否)")
    if answer == '1':
        return True
    else:
        return False

class ApproxGame(object):
    """
    Game for mental mathmatics in approximate numbers
    """
    def __init__(self):
        levels = {
        # id: (next_level, precision, round_count, recipe, *args)
        1: (2, 0, 1, 'guess', range(0, 100), 0, 100),
        2: (3, 0, 10, 'add', range(1, 10), range(0, 10)),
        6: (7, 0, 10, 'sub', range(1, 10), range(0, 10)),
        4: (5, 0, 5, 'add', 10, range(1, 10)),
        5: (6, 0, 10, 'add', range(10, 100, 10), range(1, 10)),
        3: (4, 0, 5, 'subf', 9, range(1, 10)),
        7: (8, 0, 10, 'subf', 99, range(11, 100)),
        8: (9, 0, 10, 'add', range(100, 1000, 100), range(10, 100, 10)),
        9: (10, 0, 10, 'add', range(100, 1000, 100), range(10, 100)),
        10: (11, 0, 10, 'subf', 999, range(100, 1000)),
        11: (12, 10, 10, 'add', range(1, 100), range(0, 100)),
        12: (13, 10, 10, 'sub', range(1, 100), range(0, 100)),
        13: (14, 0, 10, 'highest', 'multiply', 
                    range(10, 99), range(1, 10)),
        14: (15, 0, 10, 'multiply', range(1, 9), range(0, 9)),
        15: (16, 10, 10, 'multiply', range(10, 99), range(0, 9)),
        16: (17, 50, 10, 'multiply', range(100, 999), range(1, 9)),
        17: (18, 100, 5, 'multiply', range(10, 99), range(10, 99)),
        18: (19, 0, 10, 'divide', range(1, 9), range(2, 9)),
        19: (1, 10, 2, 'divide', range(10, 99), range(2, 9)),
        }

        self._levels = levels
        
    def play_level(self, level_id, print_method, input_func):
        level = self._levels[level_id]
        next_level, precision, round_count = level[0:3]
        recipe_args = level[3:]
        print_method(new_level_greet(level_id, precision))
        counter = 0
        while counter < round_count:
            question, answer = generate_question(*recipe_args)
            correctness = play_question(question, answer, 
                        precision, direct_print, input_func)
            if correctness:
                counter += 1
            else:
                return None
        else: 
            return next_level

    def run(self, level_id=1):
        while True:
            level_id = self.play_level(level_id, direct_print, input)
            if level_id is None:
                direct_print("=======\n遊戲結束\n=======\n")
                break

if __name__ == '__main__':
    game = ApproxGame()
    game.run()